src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/BinaryArrayRecord.cs

```diff
@@ -0,0 +1,236 @@
+                    {
+                        ThrowHelper.ThrowArrayContainedNulls();
+                    }
+
+                    int nullCount = ((NullsRecord)item).NullCount;
+                    do
+                    {
+                        array.SetValue(null, resultIndex++);
+                        nullCount--;
+                    }
+                    while (nullCount > 0);
+                    break;
+                default:
+                    array.SetValue(record.GetValue(), resultIndex++);
+                    break;
+            }
+        }
+
+        return array;
+    }
+
+    internal static ArrayRecord Parse(BinaryReader reader, RecordMap recordMap, PayloadOptions options)
+    {
+        int objectId = reader.ReadInt32();
+        BinaryArrayType arrayType = reader.ReadArrayType();
+        int rank = reader.ReadInt32();
+
+        bool isRectangular = arrayType is BinaryArrayType.Rectangular or BinaryArrayType.RectangularOffset;
+
+        if (rank < 1 || rank > 32
+            || (rank != 1 && !isRectangular)
+            || (rank == 1 && isRectangular))
+        {
+            ThrowHelper.ThrowInvalidValue(rank);
+        }
+
+        int[] lengths = new int[rank]; // adversary-controlled, but acceptable since upper limit of 32
+        long totalElementCount = 1;
+        for (int i = 0; i < lengths.Length; i++)
+        {
+            lengths[i] = ArrayInfo.ParseValidArrayLength(reader);
+            totalElementCount *= lengths[i];
+
+            if (totalElementCount > uint.MaxValue)
+            {
+                ThrowHelper.ThrowInvalidValue(lengths[i]); // max array size exceeded
+            }
+        }
+
+        int[] offsets = new int[rank]; // zero-init; adversary-controlled, but acceptable since upper limit of 32
+        bool hasCustomOffset = false;
+        if (arrayType is BinaryArrayType.SingleOffset or BinaryArrayType.JaggedOffset or BinaryArrayType.RectangularOffset)
+        {
+            for (int i = 0; i < offsets.Length; i++)
+            {
+                int offset = reader.ReadInt32();
+
+                if (offset < 0)
+                {
+                    ThrowHelper.ThrowInvalidValue(offset);
+                }
+                else if (offset > 0)
+                {
+                    hasCustomOffset = true;
+
+                    long maxIndex = lengths[i] + offset;
+                    if (maxIndex > int.MaxValue)
+                    {
+                        ThrowHelper.ThrowInvalidValue(maxIndex);
+                    }
+                }
+
+                offsets[i] = offset;
+            }
+        }
+
+        MemberTypeInfo memberTypeInfo = MemberTypeInfo.Parse(reader, 1, options, recordMap);
+        ArrayInfo arrayInfo = new(objectId, (int)totalElementCount, arrayType, rank);
+
+        if (isRectangular || hasCustomOffset)
+        {
+            return RectangularOrCustomOffsetArrayRecord.Create(arrayInfo, memberTypeInfo, lengths, offsets);
+        }
+
```

Comment:
I don't see anything reading the AdditionalTypeInfo values?

Reply:
> I don't see anything reading the AdditionalTypeInfo values?

Did you mean `AdditionalInfo`? It's used in a few places like this:

https://github.com/dotnet/runtime/blob/a7134ef7fdd5d7f05662f1f68c24204e727f2530/src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/RectangularOrCustomOffsetArrayRecord.cs#L157-L158

Reply:
No, `AdditionalTypeInfo`.

[MS-NRBF] 2.4.3.1 (BinaryArray):

> `TypeEnum` (1 byte): A BinaryTypeEnum value that identifies the Remoting Type of the Array item.
>
>`AdditionalTypeInfo` (variable): Information about the Remoting Type of the Array item in addition
to the information provided in the TypeEnum field. For the BinaryTypeEnum values of Primitive,
SystemClass, Class, or PrimitiveArray, this field contains additional information about the ...

Reply:
This part is handled by `MemberTypeInfo` here:

https://github.com/dotnet/runtime/blob/a7134ef7fdd5d7f05662f1f68c24204e727f2530/src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/MemberTypeInfo.cs#L48-L54